### DISCUSSION TITLE
Poetry-core required?

### DISCUSSION THREAD

**Discussion by cawoodjm** [python-poetry/poetry#4917] [Q&A]
Greetings everyone, 
Before I commit to logging an issue, I thought I might share my issue with the community to see if there is a solution. 

```
Poetry version 1.1.11
Python 3.8.x
OS Distro Ubuntu 20.04 LTS (Poetry framework not installed)

Tested project on MacOS BigSur 11.6 (Poetry framework installed)
```

I use poetry in the development of a private project. It works great for dependency management and building a project for distribution to a public or private space. The below .toml file is a likeness of the one used in the building my project, minus the few redacted things for privacy. Now this builds the project with ease and gives me a .whl or tarball file for distribution.

```
[tool.poetry]
name = "myproject"
version = "0.3.1"

[tool.poetry.dependencies]
python = "^3.8"
pandas = "*"
click = "*"
SQLAlchemy = "*"
poetry-core = "*"

[[tool.poetry.source]]
name = "redacted"
url = "redacted"
secondary = true

[build-system]
requires = ["poetry-core>=1.0.0"]
build-backend = "poetry.core.masonry.api"

[tool.poetry.plugins."console_scripts"]
run_function = "myproject.function:main"
```

When I setup a venv project on my linux machine and pipenv install myproject I get an error that seems odd. 

```
Collecting myproject==0.3.1
  Using cached https://repos.private-repo/myproject-0.3.1.tar.gz
  Installing build dependencies: started
  Installing build dependencies: finished with status 'error'

  ERROR: Command errored out with exit status 1:
   command: /redacted for privacy --no-warn-script-location --no-binary :none: --only-binary :none: -i https://repos.private-repo -- 'poetry-core>=1.0.0'
       cwd: None
  Complete output (3 lines):
  Looking in indexes: https://repos.private-repo/python
  ERROR: Could not find a version that satisfies the requirement poetry-core>=1.0.0 (from versions: none)
  ERROR: No matching distribution found for poetry-core>=1.0.0
```

Why would it be looking for indexes related to poetry-core in my private repo not trying to install from pypi? Now I assume this could be due to the fact that Poetry is not pip installed at the OS level using the curl command [Poetry docs](https://python-poetry.org/docs/#osx--linux--bashonwindows-install-instructions) provides. I thought thought adding poetry-core as a dependency in my core project pyproject.toml (seen above) would resolve that dependency, but it did not. So does Poetry have to be pip installed at the OS level for any project that uses ["poetry-core>=1.0.0"] as part of the build process?

**Comment by abn** [comment]
What you are seeing here is pip attempting to install your package from its sdist. When this happens, 

- pip creates an isolated build environment for PEP517 build
- the build environment requires the PEP517 backend and any other build requirements for the build system, which in this case is poetry-core

This step is where, if your pip global config uses a private index, poetry-core is searched for.

**Reply by cawoodjm** [reply]
Thank you for describing what is happening. What I am confused is if poetry-core is listed in the dependencies when I build the sdist should it not first look on pypi first before the private repo. I do not get why this issue does not impact with click, pandas, or sqlalchemy?

[tool.poetry.dependencies]
python = "^3.8"
pandas = "*"
click = "*"
SQLAlchemy = "*"
poetry-core = "*"

**Reply by abn** [reply]
I am not sure why `pipenv` would respect poetry configuration, as `tool.poetry.source` configuration is only respected by Poetry. When you are using `pipenv` what I suspect is happening is that it uses its index configuration (same as that in `~/.pip/pip.conf` or configured via env vars) when it creates a PEP517 "build" environment for your package. The `poetry-core` in question here is not your project's dependency, but rather the `build-system.requires` entry of `poetry-core>=1.0.0`. And it would seem your `pip` ins configured to use your private repository.

You can in theory avoid this issue by or verify this by, performing `poetry build` then using the wheel from `dist/` like `pip install ./dist/*.whl` in a new environmnt. What I suspect will happen is that pip will now determine your proejct dependencies and try to fetch then all from the private repository. Which again is not what you want unless your private repo is also a PyPI proxy.
